Updating Java Design Guidelines to match latest spec (#2124)
Azure/azure-sdk: docs/java/introduction.md

---
title: "Java Azure SDK Design Guidelines"
keywords: guidelines java
permalink: java_introduction.html
folder: java
sidebar: general_sidebar
---

## Introduction

The following document describes Java specific guidelines for designing Azure SDK client libraries. These guidelines also expand on and simplify language-independent [General Azure SDK Guidelines][general-guidelines]. More specific guidelines take precedence over more general guidelines.

The Java guidelines are for the benefit of client library designers targeting service applications written in Java.  If you are a client library designer that is targeting Android mobile apps, refer to the [Android Guidelines](android_introduction.html) instead.

> ## Design Principles {#java-principles}

Author: These are used for the anchor links. It's fine to have as-is I believe.

**Idiomatic**

* The SDK should follow the general design guidelines and conventions of Java. It should feel natural to a Java developer.
* We embrace the ecosystem with its strengths and its flaws.
* We work with the ecosystem to improve it for all developers.
* Azure SDK libraries version just like standard Java libraries.

> We are not trying to fix bad parts of the language ecosystem; we embrace the ecosystem with its strengths and its flaws.

**Consistent**

* The Azure SDK feels like a single product of a single team, not a set of Maven libraries.
* Users learn common concepts once; apply the knowledge across all SDK components.
* All differences from the guidelines must have good reasons.

**Approachable**

* Small number of steps to get started; power knobs for advanced users
* Small number of concepts; small number of types; small number of members
* Approachable by our users, not by engineers designing the SDK components
* Easy to find great _getting started_ guides and samples
* Easy to acquire

> **Dependable**

Author: .NET decided to remove it - but I'll leave it up to @annelo-msft to determine if that was correct or a mistake. I'll follow suit in Java.

> **Dependable**

Reviewer: Different language guidelines have different subsets of the principles - reconciling these is outside the scope of MQ, but I've added it to the list of post-MQ suggestions.  https://github.com/Azure/azure-sdk/issues/2118

For now, I'm happy for you to make the judgment call, @JonathanGiles, depending on how strong of a focus you feel this is for Java.

### Support for non-HTTP Protocols

Currently, this document describes guidelines for client libraries exposing HTTP services, with a small amount of guidance for AMQP-based services. If your service is not HTTP-based, please contact the [Azure SDK Architecture Board](https://azure.github.io/azure-sdk/policies_reviewprocess.html) for guidance.

## Azure SDK API Design {#java-api}

Azure services are exposed to Java developers as one or more *service client* types and a set of *supporting types*.

### Service Client

Service clients are the main starting points for developers calling Azure services with the Azure SDK. Each client library should have at least one client in its main namespace, so it’s easy to discover. The guidelines in this section describe patterns for the design of a service client. Because in Java both synchronous and asynchronous service clients are required, the sections below are organized into general service client guidance, followed by sync- and async-specific guidance.

{% include requirement/MUST id="java-service-client-name" %} name service client types with the _Client_ suffix (for example, `ConfigurationClient`).

> {% include requirement/MUST id="java-service-client-annotation" %} annotate all service clients with the `@ServiceClient` annotation.

Author: Done.

{% include requirement/MUST id="java-service-client-namespace" %} place service client types that the consumer is most likely to interact with in the root package of the client library (for example, `com.azure.<group>.servicebus`).  Specialized service clients should be placed in sub-packages.

> {% include requirement/MUST id="java-service-client-immutable" %} ensure that all service client classes are immutable upon instantiation.

Author: Done

    }

    // A non-paginated sync list API (refer to pagination section for more details)
    public IterableStream<<model>> list<service_operation>(<parameters>) {
        // ...
    }

    // A paginated sync list API (refer to pagination section for more details)
    public PagedIterable<<model>> list<service_operation>(<parameters>) {
        // ...
    }

    // other members
    …
}
```

Refer to the [ConfigurationClient class] for a fully built-out example of how a sync client should be constructed.

#### Async Service Clients

{% include requirement/MUST id="java-async-client-name" %} offer an async service client named `<ServiceName>AsyncClient`. More than one service client may be offered for a single service. An example of an async client is shown below:

```java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.<group>.<service_name>;

@ServiceClient(
    builder = <service_name>ClientBuilder.class,
    serviceInterfaces = <service_name>Service.class,
    isAsync = true)
public final class <service_name>AsyncClient {

    // package-private constructors only - all instantiation is done with builders

>     <service_name>Client(<parameters>, HttpPipeline pipeline) {

Author: Done

>         super(pipeline);

Author: Not sure - the approach changed and this didn't keep up. Removing...

    }

    // service methods...

    // A single response API
    public Mono<Response<<model>>> <service_operation>(<parameters>) {
        ...
    }

    // A paginated response API
    public PagedFlux<<model>> list<service_operation>(<parameters>) {
        ...
    }

    // other members
    ...
}
```

Refer to the [ConfigurationAsyncClient class] for a fully built-out example of how an async client should be constructed.

{% include requirement/MUST id="java-async-framework" %} use [Project Reactor] to provide consumers with a high-quality async API.

{% include requirement/MUSTNOT id="java-async-other-frameworks" %} use any other async APIs, such as `CompletableFuture` or [RxJava].

> {% include requirement/MUSTNOT id="java-async-streaming" %} write custom APIs for streaming or async operations. Make use of the existing functionality offered in the Azure core library. Discuss proposed changes to the Azure core library with the [Architecture Board]. Refer to the [Azure Core Types](#using-azure-core-types) section for more information.

Author: Yes, I do say that elsewhere that subtypes are allowed: "Subclasses of these types are acceptable as return types too, so long as the naming convention generally follows the pattern `<serviceName>PagedIterable` or `<operation>PagedIterable`."

#### Service Client Creation

> {% include requirement/MUSTNOT id="java-service-client-constructors" %} provide any `public` or `protected` constructors in the service client, except where necessary to support mock testing. Keep visibility to a minimum by using package-private constructors that may only be called by types in the same package, and then enable instantiation of the service client through the [service client builder](#service-client-creation).

Author: Will clarify

> {% include requirement/MUSTNOT id="java-service-client-constructors" %} provide any `public` or `protected` constructors in the service client, except where necessary to support mock testing. Keep visibility to a minimum by using package-private constructors that may only be called by types in the same package, and then enable instantiation of the service client through the [service client builder](#service-client-creation).

Author: Fixed.

{% include requirement/MUST id="java-service-client-builder-consistency" %} ensure consistency across all HTTP-based client libraries, by using the following names for client builder fluent API:

| Name                 | Intent                                                                                                 |
|----------------------|--------------------------------------------------------------------------------------------------------|
| `addPolicy`          | Adds a policy to the set of existing policies (assumes no custom pipeline is set).                     |
| `buildAsyncClient`   | Creates a new **async** client on each call.                                                           |
| `buildClient`        | Creates a new **sync** client on each call.                                                            |
| `clientOptions`      | Allows the user to set a variety of client-related options, such as user-agent string, headers, etc.   |
| `configuration`      | Sets the configuration store that is used during construction of the service client.                   |

> | `connectionString`   | Sets the connection string to use for (only applicable if the Azure portal offers it for the service). |

Author: Fixed.

{% include requirement/MUST id="java-service-client-builder-consistency-amqp" %} ensure consistency across all AMQP-based client libraries, by using the following names for client builder fluent API:

| Name                       | Intent                                                                                                 |
|----------------------------|--------------------------------------------------------------------------------------------------------|
| `build<Type>AsyncClient`   | Creates a new **async** client on each call.                                                           |
| `build<Type>Client`        | Creates a new **sync** client on each call.                                                            |
| `configuration`            | Sets the configuration store that is used during construction of the service client.                   |
| `credential`               | Sets the credential to use when authenticating AMQP requests.                                          |

> | `connectionString`         | Sets the connection string to use for (only applicable if the Azure portal offers it for the service). |

Author: Fixed

| `transportType`            | Sets the preferred transport type to AMQP or Web Sockets that the client should use.                   |
| `retry`                    | Sets the retry policy to use (using the `AmqpRetryOptions` type).                                      |
| `proxyOptions`             | Sets the proxy connection settings.                                                                    |
| `serviceVersion`           | Sets the [service version](#versioning) to use. This must be a type implementing `ServiceVersion`.     |

{% include requirement/MUST id="java-service-client-builder-precedence" %} use the following precedence rules for builder arguments:

1. If the user sets a non-null `pipeline`, all other settings related to construction and configuration of a pipeline are ignored. The provided pipeline is used as-is.

> 2. If the user sets a `connectionString` and a `credential`, the `credential` will take precedence and the `connectionString` must be ignored.

Author: Yeah, let me tighten up that guidance, as my preference is to throw in the build method too.

```java
public enum ConfigurationServiceVersion implements ServiceVersion {
    V1_0("1.0");

    private final String version;

    ConfigurationServiceVersion(String version) {
        this.version = version;
    }

    @Override
    public String getVersion() {
        return this.version;
    }

    /**
     * Gets the latest service version supported by this client library
     *
     * @return the latest {@link ConfigurationServiceVersion}
     */
    public static ConfigurationServiceVersion getLatest() {
        return V1_0;
    }
}
```

This can then be called by the developer as such:

```java
public class UserApplication {
    public static void main(String args[]) {
        ConfigurationClient client = new ConfigurationClientBuilder()
            .credential(<tokenCredential>)
            .endpoint("<endpoint>")
            .serviceVersion(ConfigurationServiceVersion.V1_0) // set the version to V1
            .buildClient();

        // calls V1 service API
        ConfigurationSetting setting = client.getConfigurationSetting("name", "label"); 
    }
}
```

> {% include note.html content="Third-party reusable libraries shouldn't change behavior without an explicit decision by the developer.  When developing libraries that are based on the Azure SDK, lock the library to a specific service version to avoid changes in behavior." %}

Author: Yep, can remove it.

#### Service Methods

Service methods are the methods on the client that invoke operations on the service.

> {% include requirement/MUST id="java-service-client-method-naming" %} use standard JavaBean naming prefixes for all getters and setters that are not service methods.

Author: Fixed

One of the Azure Core types is `com.azure.core.util.Context`, which acts as an append-only key-value map, and which by default is empty. The `Context` allows end users of the API to modify the outgoing requests to Azure on a per-method call basis, for example to enable distributed tracing.

{% include requirement/MUST id="java-service-client-context" %} provide an overload method that takes a `com.azure.core.util.Context` argument for each service operation **in sync clients only**. The `Context` argument must be the last argument into the service method (except where `varargs` are used). If a service method has multiple overloads, only the 'maximal' overloads need to have the `Context` argument. A maximal overload is one that has a full set of arguments.  It may not be necessary to offer a 'Context overload' in all cases.  We prefer a minimal API surface, but `Context` must always be supported.

```java
getFoo()
getFoo(x)
getFoo(x, y)
getFoo(x, y, z) // maximal overload
getFoo(a)       // maximal overload

> // this will result in the following two methods being required:

Author: Fixed

getFoo(x, y, z, Context)
getFoo(a, Context)
```

{% include requirement/MUSTNOT id="java-service-client-context-async" %} include overloads that take `Context` in async clients.  Async clients use the [subscriber context built into Reactor Flux and Mono APIs][reactor-context].

##### Cancellation

> {% include requirement/MUSTNOT id="java-no-cancellation" %} provide any API that accepts a cancellation token, in both sync and async clients. Cancellation isn't a common pattern in Java. Developers who use our client libraries, and who need to cancel requests, should use the async API instead, where they can unsubscribe from a publisher to cancel the request.

Author: Are you saying to state this as a guidance to "try your best to cancel when a user unsubscribes"?

The *logical entity* is a protocol neutral representation of a response. The logical entity may combine data from headers, body, and the status line. For example, you may expose an `ETag` header as a property on the logical entity.  `Response<T>` is the 'complete response'. It contains HTTP headers, status code, and the `T` object (a deserialized object created from the response body). The `T` object would be the 'logical entity'.

{% include requirement/MUST id="java-response-logical-entity" %} return the logical entity (i.e. the `T`) for all **synchronous** service methods.

{% include requirement/MUST id="java-response-async-logical-entity" %} return the logical entity (i.e. the `T`) wrapped inside a `Mono` for all **asynchronous** service methods that make network requests.

Return `Response<T>` on the maximal overload for a service method with `WithResponse` appended to the name.  For example:

```java
Foo foo = client.getFoo(a);
Foo foo = client.getFoo(a, b);
Foo foo = client.getFoo(a, b, c, context); // This is the maximal overload

> Response<Foo> response = client.getFooWithResponse(a, b, c, context);

Author: Done

```

For methods that combine multiple requests into a single call:

> {% include requirement/MUSTNOT id="java-response-headers" %} return headers and other per-request metadata unless it's obvious which specific HTTP request the methods return value corresponds to.

Author: This is for methods that combine multiple requests into a single call. I believe it is to avoid having Iterable<Response<T>> APIs, so it is your second question that is correct.

        return access;
    }

    public CreateBlobOptions setMetadata(Map<String, String> metadata) {
        this.metadata = metadata;
        return this;
    }

    public Map<String, String> getMetadata() {
        return metadata;
    }
}
```

{% include requirement/MUST id="java-params-complex-naming" %} name the _options_ type after the name of the service method it is used for, such that the type is named `<operation>Options`. For example, above the method was `createBlob`, and so the _options_ type was named `CreateBlobOptions`.

{% include requirement/MUST id="java-params-complex" %} use the _options_ parameter pattern for complex service methods.

{% include requirement/MAY id="java-params-complex-growth" %} use the _options_ parameter pattern for simple service methods that you expect to `grow` in the future.

{% include requirement/MAY id="java-params-complex-overloads" %} add simple overloads of methods using the _options_ parameter pattern.

If in common scenarios, users are likely to pass just a small subset of what the _options_ parameter represents, consider adding an overload with a parameter list representing just this subset.

> {% include requirement/MUSTNOT id="java-params-complex-overloads" %} introduce method overloads that take a subset of the parameters as well as the _options_ parameter. _Option_ parameters must be the only argument to a method, apart from the `Context` type, which must continue to be outside the _options_ type.

Author: I'll clarify

#### Methods Invoking Long-Running Operations

Some service operations, known as _Long Running Operations_ or _LROs_ take a long time (up to hours or days). Such operations do not return their result immediately, but rather are started, their progress is polled, and finally the result of the operation is retrieved.

The [azure-core](#using-azure-core-types) library exposes types that make supporting LROs simpler, most notably `SyncPoller<T>` and `PollerFlux<T>`, for synchronous and asynchronous clients, respectively. These classes represent a LRO and support operations for polling and waiting for status changes, and retrieving the final operation result.

{% include requirement/MUST id="java-lro-poller-class" %} use the `SyncPoller` and `PollerFlux` to represent long-running operations. The long-running operation API pattern is:

```java
// sync client

> public class <service_name>Client {

Author: Done

    // SyncPoller<T, U> is a type in azure core
    // T is the type of long-running operation poll response value
    // U is the type of the final result of long-running operation
    public SyncPoller<T, U> begin<operation_name>(<parameters>) {
        PollerFlux<> asyncPoller = asyncClient.begin<operation_name>(<parameters>);
        return asyncPoller.getSyncPoller();
    }
}
```

```java
// Async client

> public class <service_name>AsyncClient {

Author: Done

    // PollerFlux<T, U> is a type in azure core
    // T is the type of long-running operation poll response value
    // U is the type of the final result of long-running operation
    public PollerFlux<T, U> begin<operation_name>(<parameters>) {
        return new PollerFlux<>(...);
    }
}
```

{% include requirement/MUST id="java-lro-prefix" %} prefix method names which return a poller with the `begin` prefix.

{% include requirement/MUST id="java-lro-continuation" %} provide a way to instantiate a poller with the serialized state of another poller to begin where it left off, for example by passing the state as a parameter to the same method which started the operation, or by directly instantiating a poller with that state.

> TODO this rehydration guidance is being written based on recent development efforts and will be updated soon

#### Conditional Request Methods

> [Conditional requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests) are normally performed using HTTP headers.  The primary usage provides headers that match the `ETag` to some known value.  The `ETag` is an opaque identifier that represents a single version of a resource. For example, adding the following header will translate to "if the record's version, specified by the `ETag`, is not the same".

Author: Done

{% include requirement/MUST id="java-namespaces-format" %} construct the package name with all lowercase letters (no camel case is allowed), without spaces, hyphens, or underscores. For example, Azure Key Vault would be in `com.azure.security.keyvault` - note that the two words 'Key' and 'Vault' are brought together to `keyvault`, instead of `keyVault`, `key_vault`, or `key-vault`.  It may further be shortened if the shortened version is well known in the community.  For example, "Azure Media Analytics" would have a compressed service name of `mediaanalytics`, and "Azure Service Bus" would become `servicebus`.

{% include requirement/MUST id="java-namespaces-package-name" %} pick a package name that allows the consumer to tie the package to the service being used. The package does **NOT** change when the branding of the product changes.  Avoid the use of marketing names that may change.

{% include requirement/MUST id="java-namespaces-approved-list" %} use the following list as the group of services:

{% include tables/data_namespaces.md %}

If the client library does not seem to fit into the group list, contact the [Architecture Board] to discuss the namespace requirements.

> {% include requirement/MUST id="java-namespaces-management" %} place the management (Azure Resource Manager) API in the `management` group.  Use the grouping `<AZURE>.management.<group>.<service>` for the namespace. Since more services require control plane APIs than data plane APIs, other namespaces may be used explicitly for control plane only.  Data plane usage is by exception only.  Additional namespaces that can be used for control plane SDKs include:

Reviewer 2: Should this now be updated to the new name `resourcemanager`?

Author: Yep

{% include tables/mgmt_namespaces.md %}

> Many `management` APIs do not have a data plane because they deal with management of the Azure account. Place the management library in the `com.azure.management` namespace.  For example, use `com.azure.management.costanalysis` instead of `com.azure.management.management.costanalysis`.

Reviewer 2: Update the namespace here as well.

Author: Done

* The information will be shared between multiple client libraries.

Let's take two examples:

1. Implementing two Cognitive Services client libraries, we find a model is required that is produced by one Cognitive Services client library and consumed by another Cognitive Services client library, or the same model is produced by two client libraries. The consumer is required to do the passing of the model in their code, or may need to compare the model produced by one client library vs. that produced by another client library. This is a good candidate for choosing a common library.

2. Two Cognitive Services client libraries throw an `ObjectNotFound` exception to indicate that an object was not detected in an image. The user might trap the exception, but otherwise will not operate on the exception. There is no linkage between the `ObjectNotFound` exception in each client library. This is not a good candidate for creation of a common library (although you may wish to place this exception in a common library if one exists for the namespace already). Instead, produce two different exceptions - one in each client library.

#### Java 9 Modules

> Java 9 and later support the notion of a module. A module *exports* certain packages, and *requires* other modules. Any package that is exported can be used by other modules, and anything that is not exported is invisible at compile and run times. This is a far stronger form of encapsulation than has existed previously for Java. For the Azure SDK for Java, a client library will be repesented as one or more modules. Two good resources to understand modules are available on [oracle.com](https://www.oracle.com/corporate/features/understanding-java-9-modules.html) and [baeldung.com](https://www.baeldung.com/java-9-modularity).

Reviewer 2: ```suggestion
Java 9 and later support the notion of a module. A module *exports* certain packages, and *requires* other modules. Any package that is exported can be used by other modules, and anything that is not exported is invisible at compile and run times. This is a far stronger form of encapsulation than has existed previously for Java. For the Azure SDK for Java, a client library will be represented as one or more modules. Two good resources to understand modules are available on [oracle.com](https://www.oracle.com/corporate/features/understanding-java-9-modules.html) and [baeldung.com](https://www.baeldung.com/java-9-modularity).
```

Author: Done

{% include requirement/MUSTNOT id="java-module-no-conditional-exports" %} conditionally `export` or `opens` packages to other modules without prior approval by the architecture board. A conditional `export` or `opens` statement takes the form `export X to Y` or `opens X to Y`.

{% include requirement/MUSTNOT id="java-module-split-packages" %} have the same package in multiple modules. That is, do not have `com.azure.storage.blob` in module `com.azure.storage.blob` and in module `com.azure.storage.blob.extras`. It is however allowed to have different packages with common parent packages split across different modules. For example, a package named `com.azure.storage` can be in one module, and `com.azure.storage.blob` can be in another.

### Versioning

{% include requirement/MUST id="java-versioning-backwards-compatibility" %} be 100% backwards compatible with older versions of the same package.

{% include requirement/MUST id="java-versioning-highest-api" %} call the highest supported service API version by default.

> {% include requirement/MUST id="java-versioning-select-api-version" %} allow the consumer to explicitly select a supported service API version when instantiating the service client, by using the service client builder with a property called `serviceVersion`. This method must take a type implementing the `ServiceVersion` interface, named specifically for the service, but as generally as possible. For example, `IdentityServiceVersion` for Identity. For a service with multiple sub-services, such as Storage, if the services all share a common versioning system, `StorageServiceVersion` would suffice. If they did not, it would be necessary to have separate `BlobServiceVersion`, `QueueServiceVersion`, and `FileServiceVersion` enums.

Reviewer 2: We know that new versions will be added. Can we make this guideline to always use ExpandableStringEnum instead of `enum`? We have already shipped some using enum but we should use ExpandableStringEnum instead for future services.

Author: For ServiceVersion we made the conscious decision to use enum, because it makes no sense to support users in specifying their own service version string.

#### Client Version Numbers {#java-versionnumbers}

Consistent version number scheme allows consumers to determine what to expect from a new version of the library.

{% include requirement/MUST id="java-version-semver" %} use _MAJOR_._MINOR_._PATCH_ format for the library version.

Use `-beta.N` suffix for beta package versions. For example, `1.0.0-beta.2`.

{% include requirement/MUST id="java-version-change-on-release" %} change the version number of the client library when **ANYTHING** changes in the client library.

{% include requirement/MUST id="java-version-patching" %} increment the patch version when fixing a bug.

{% include requirement/MUSTNOT id="java-version-features-in-patch" %} include new APIs in a patch release.

{% include requirement/MUST id="java-version-add-feature" %} increment the major or minor version when adding support for a service API version.

{% include requirement/MUST id="java-version-add-api" %} increment the major or minor version when adding a new method to the public API.

{% include requirement/SHOULD id="java-version-major-changes" %} increment the major version when making large feature changes.

> {% include requirement/MUST id="java-version-change-on-release" %} select a version number greater than the highest version number of any other released Track 1 packages for the service.

Reviewer 2: This must take into consideration the highest major version of Track 1 packages in *all languages* and then use the number greater than that.

### Dependencies

> Dependencies bring in many considerations that are often easily avoided by avoiding the 
> dependency. 

Reviewer 2: ```suggestion
Dependencies bring in many considerations that are often easily avoided by avoiding the dependency. 
```

Author: Fixed

{% include requirement/MUSTNOT id="java-dependencies-concrete" %} depend on concrete logging, dependency injection, or configuration technologies (except as implemented in the `com.azure.core` library).  The client library will be used in applications that might be using the logging, DI, and configuration technologies of their choice.

### Native code

Native code plugins cause compatibility issues and require additional scrutiny. Certain languages compile to a machine-native format (for example, C or C++), whereas most modern languages opt to compile to an intermediary format to aid in cross-platform support.

{% include requirement/MUSTNOT id="java-no-native-code" %} write platform-specific / native code.

### Documentation

{% include requirement/MUST id="java-javadoc-build" %} ensure that anybody can clone the repo containing the client library and execute `mvn javadoc:javadoc` to generate the full and complete JavaDoc output for the code, without any need for additional processing steps.

> {% include requirement/MUST id="java-javadoc-samples" %} include code samples in all class-level JavaDoc, and in relevant method-level JavaDoc.

Reviewer 2: Add a guideline to include documentation for all params, return type and known exceptions that can be thrown by a method.

{% include requirement/MUSTNOT id="java-javadoc-hard-coding" %} hard-code the sample within the JavaDoc (where it may become stale). Put code samples in `/src/samples/java` and use the available tooling to reference them.

{% include requirement/MUST id="java-javadoc-naming-samples" %} follow the naming convention outlined below for naming samples tags:

 * If a new instance of the class is created through build() method of a builder or through constructor: `<packagename>.<classname>.instantiation`
 * For other methods in the class: `<packagename>.<classname>.<methodName>`
 * For overloaded methods, or methods with arguments: `<packagename>.<classname>.<methodName>#<argType1>-<argType2>`
 * Camel casing for the method name and argument types is valid, but not required.

## Repository Guidelines

> ### Documentation

Reviewer 2: There's already a documentation section above. Should this be removed?

Author: I believe the intent is to separate out the documentation that users write for code (i.e. javadoc) versus other documentation that goes into the repo.

> We are not trying to fix bad parts of the language ecosystem; we embrace the ecosystem with its strengths and its flaws.

**Consistent**

* The Azure SDK feels like a single product of a single team, not a set of Maven libraries.
* Users learn common concepts once; apply the knowledge across all SDK components.
* All differences from the guidelines must have good reasons.

**Approachable**

* Small number of steps to get started; power knobs for advanced users
* Small number of concepts; small number of types; small number of members
* Approachable by our users, not by engineers designing the SDK components
* Easy to find great _getting started_ guides and samples
* Easy to acquire

**Dependable**

* 100% backward compatible
* Great logging, tracing, and error messages
* Predictable support lifecycle, feature coverage, and quality

### General Guidelines

{% include requirement/MUST id="java-general-follow-general-guidelines" %} follow the [General Azure SDK Guidelines].

{% include requirement/MUST id="java-general-repository" %} locate all source code in the [azure/azure-sdk-for-java] GitHub repository.

### Support for non-HTTP Protocols

> Currently, this document describes guidelines for client libraries exposing HTTP services, with a small amount of guidance for AMQP-based services. If your service is not HTTP-based, please contact the Azure SDK Architecture Board for guidance.

Reviewer: You can find public contact info at this link: https://azure.github.io/azure-sdk/policies_reviewprocess.html.  Maybe link to this as follows?

```suggestion
Currently, this document describes guidelines for client libraries exposing HTTP services, with a small amount of guidance for AMQP-based services. If your service is not HTTP-based, please contact the [Azure SDK Architecture Board](https://azure.github.io/azure-sdk/policies_reviewprocess.html) for guidance.
```

Reviewer: Gah, I just saw the remaining TODO below!

Author: Fixed

> TODO: Link to current arch board contact info.

## Azure SDK API Design {#java-api}

Azure services are exposed to Java developers as one or more *service client* types and a set of *supporting types*.

### Service Client

> Service clients are the main starting points for developers calling Azure services with the Azure SDK. Each client library should have at least one client in its main namespace, so it’s easy to discover. The guidelines in this section describe patterns for the design of a service client, however, because in Java the guidelines below require both a synchronous and an asynchronous service client, the sections below are split into general guidance, followed by sync- and async-specific guidance.

Reviewer: Suggest breaking up this last sentence to make it easier to digest.

```suggestion
Service clients are the main starting points for developers calling Azure services with the Azure SDK. Each client library should have at least one client in its main namespace, so it’s easy to discover. The guidelines in this section describe patterns for the design of a service client.  Because in Java both synchronous and asynchronous service clients are required, the sections below are organized into general service client guidance, followed by sync- and async-specific guidance.
```

Author: Fixed

{% include requirement/MUST id="java-service-client-annotation" %} annotate all service clients with the `@ServiceClient` annotation.

{% include requirement/MUST id="java-service-client-namespace" %} place service client types that the consumer is most likely to interact with in the root package of the client library (for example, `com.azure.<group>.servicebus`). Specialized service clients should be placed in sub-packages.

> {% include requirement/MUST id="java-service-client-immutable" %} ensure that all service client classes are immutable and stateless upon instantiation.

Reviewer: Just curious - are the endpoint and credential not considered client state?

Refer to the [ConfigurationAsyncClient class] for a fully built-out example of how an async client should be constructed.

{% include requirement/MUST id="java-async-framework" %} use [Project Reactor] to provide consumers with a high-quality async API.

{% include requirement/MUSTNOT id="java-async-other-frameworks" %} use any other async APIs, such as `CompletableFuture` or [RxJava].

{% include requirement/MUSTNOT id="java-async-streaming" %} write custom APIs for streaming or async operations. Make use of the existing functionality offered in the Azure core library. Discuss proposed changes to the Azure core library with the [Architecture Board]. Refer to the [Azure Core Types](#using-azure-core-types) section for more information.

#### Service Client Creation

> {% include requirement/MUSTNOT id="java-service-client-constructors" %} provide any `public` or `protected` constructors in the service client, except where necessary to support mock testing. Keep visibility to a minimum. Instantiation of the service client must always be through a service client builder.

Reviewer: > Keep visibility to a minimum.

This feels like it could be interpreted in a variety of ways.  Is there a more specific rule you could give?  How do we minimize visibility of constructors in Java?  (Maybe a code sample would help here)

Author: Fixed

}
```

```java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
package com.azure.data.appconfiguration;

> // concrete example of a builder

Reviewer: Thank you for this - now I know what this famous builder pattern I've heard of looks like! 😊

@ServiceClientBuilder(serviceClients = {ConfigurationAsyncClient.class, ConfigurationClient.class})
public final class ConfigurationClientBuilder {
    private String endpoint;
    private TokenCredential tokenCredential;
    private ConfigurationServiceVersion version = ConfigurationServiceVersion.getLatest();
    // other fields and its setters are omitted for brevity

>     // public constructor - this is the only available front door to creating a service client instance

Reviewer: Is this the one for mocking?  Might be worth noting that in the comment in case someone missed it in the guideline.

Author: This is the builder, so it is allowed because this is how people instantiate the service client. Will clarify.

Author: Fixed

{% include requirement/MUST id="java-service-client-builder-state" %} throw an `IllegalStateException` from the builder method when it receives mutually exclusive arguments.  The consumer is over-specifying builder arguments, some of which will necessarily be ignored. The error message in the exception must clearly outline the issue.

{% include requirement/MUST id="java-client-construction" %} allow the consumer to construct a service client with the minimal information needed to connect and [authenticate](#authentication) to the service.

> {% include requirement/MUST id="java-service-client-builder-validity" %} ensure the builder will instantiate a service client into a valid state.  Throw an `IllegalStateException` when the user calls the `build*()` methods with a configuration that is incomplete or invalid.

Reviewer: Really nice clarification! ❤

##### Service Versions

{% include requirement/MUST id="java-versioning-highest-api" %} call the highest supported service API version by default, and ensure this is clearly documented.

{% include requirement/MUST id="java-versioning-select-api-version" %} allow the consumer to explicitly select a supported service API version when instantiating the service client, as shown above in the [service client creation](#service-client-creation) section.

Use a builder parameter called `serviceVersion` on the client builder type (as [specified above](#service-client-creation)).

> For example, the following is a code snippet from the `ConfigurationServiceVersion`:

Reviewer: Will this always need to be a subtype of `ServiceVersion`?  If so, it might be worth calling out that that is a requirement and not just an illustration from the example.

Author: Done

getFoo(a, Context)
```

{% include requirement/MUSTNOT id="java-service-client-context-async" %} include overloads that take `Context` in async clients.  Async clients use the [subscriber context built into Reactor Flux and Mono APIs][reactor-context].

##### Cancellation

> {% include requirement/MUSTNOT id="java-no-cancellation" %} provide any API that accepts a cancellation token, in both sync and async clients. Cancellation isn't a common pattern in Java.  Developers who need to cancel requests should use the async API instead, where they can unsubscribe from a publisher to cancel the request.

Reviewer: Does this warrant a code sample?  Will an API designer know what to do given this text?

Author: I've clarified that this is an instruction for end developers rather than the client library developer:

{% include requirement/MUSTNOT id="java-no-cancellation" %} provide any API that accepts a cancellation token, in both sync and async clients. Cancellation isn't a common pattern in Java. Developers who use our client libraries, and who need to cancel requests, should use the async API instead, where they can unsubscribe from a publisher to cancel the request.

##### Return Types

Requests to the service fall into two basic groups: methods that make a single logical request, and methods that make a deterministic sequence of requests.  An example of a *single logical request* is a request that may be retried inside the operation.  An example of a *deterministic sequence of requests* is a paged operation.

The *logical entity* is a protocol neutral representation of a response. The logical entity may combine data from headers, body, and the status line. For example, you may expose an `ETag` header as a property on the logical entity.  `Response<T>` is the 'complete response'. It contains HTTP headers, status code, and the `T` object (a deserialized object created from the response body). The `T` object would be the 'logical entity'.

> {% include requirement/MUST id="java-response-logical-entity" %} return the logical entity (i.e. the `T`) for all **synchronous** service methods.

Reviewer: Thank you! :)

#### Methods Returning Collections (Paging)

> Many Azure REST APIs return collections of data in batches or pages. A client library will expose such APIs as special enumerable types `PagedIterable<T>` or `PagedFlux<T>`, for synchronous and asynchronous APIs, respectively.

Reviewer: ❤

{% include requirement/MUST id="java-api-file-paths" %} represent file paths using the Java `java.nio.file.Path` type. Do not use String or the older `java.io.File` type.

{% include requirement/MUST id="java-wrap-primitives" %} wrap primitive types where appropriate to represent a meaningful domain entity even if the model type contains a single field. For example, a phone number is just a string, but creating a new type to wrap primitive `String` type can be more informative and represents a domain concept. It may also provide stronger guarantees and validation than just the primitive type.

```java
public final class PhoneNumber {
    private String phoneNumber;

    public PhoneNumber setPhoneNumber(String phoneNumber) {
        ...
    }

    public String getPhoneNumber() {
        ...
    } 
}
```

### Exceptions

> Error handling is an important aspect of implementing a client library. It is the primary method by which problems are communicated to the consumer. There are two methods by which errors are reported to the consumer. Either the method throws an exception, or the method returns an error code (or value) as its return value, which the consumer must then check. In this section we refer to "producing an error" to mean throwing an exception, and "an error" to be the error value or exception object.

Reviewer: I think when we discussed this you had said Java always returns an exception, never an error code.  If this is true, consider removing the error code text from this part of the discussion?

Author: I could have sworn I did delete this, not too sure if I missed a part of it or it crept back in somehow. Will remove (again).

Author: Done

- `com.azure.mixedreality.kinect` (the grouping is not in the approved list)

### Support for Mocking

All client libraries must support mocking to enable non-live testing of service clients by customers.

Below is an example of writing a mock unit test using the [Mockito framework](https://site.mockito.org/). For more details on using Mockito in the context of the Azure SDK for Java, refer to the [unit testing](https://github.com/Azure/azure-sdk-for-java/wiki/Unit-Testing) wiki documentation.

```java
public class UserPreferencesTest {
    @Test
    public void getThemeTest() {
        // create a mock instance of client
        ConfigurationClient configurationClient = mock(ConfigurationClient.class);

        // mock the client response
        when(configurationClient.getSetting("theme", null)).thenReturn(new ConfigurationSetting().setValue("light"));

        // wire the mock client to UserPreferences
        UserPreferences userPreferences = new UserPreferences(configurationClient);
        
        // assert the client response
        assertEquals(Theme.LIGHT, userPreferences.getTheme());
    }
}
```

> {% include requirement/MUST id="java-mocking" %} support mocking to enable non-live testing of service clients by customers.

Reviewer: Should mocking only be supported for service clients, or are there other Azure.Core types that should support mocking as well?

Author: Clarified

## Azure SDK Library Design

### Packaging

> All client libraries for Java standardize on the Maven build tooling for build and dependency management. This section details the standard configuration that must be used in all client libraries.

Reviewer: Is it worth adding a #### Maven header here or no?

Author: Done

Combined operations require knowledge of additional operations that might be outside their current focus. The developer must first understand the code surrounding the operation they're working on, and can't copy and paste the code sample into their project.

## Java API Best Practices

> This section introduces guidelines for fundamental Java development design decisions that are used throughout the Azure SDK for Java. For further guidance, consider reviewing presentations and documentation on [Java API design best practices](https://www.jonathangiles.net/java-api-design-best-practices.html).

Reviewer: ❤
